docs/proposals/network-policy.md

```diff
@@ -0,0 +1,174 @@
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NetworkPolicy
+
+## Abstract
+
+A proposal for implementing a new resource - NetworkPolicy - which
+will enable definition of ingress policies for selections of pods.
+
+The design for this proposal has been created by, and discussed
+extensively within the Kubernetes networking SIG.  It has been implemented
+and tested using Kubernetes API extensions by various networking solutions already.
+
+In this design, users can create various NetworkPolicy objects which select groups of pods and
+define how those pods should be allowed to communicate with each other.  The
+implementation of that policy at the network layer is left up to the
+chosen networking solution.
+
+> Note that this proposal does not yet include egress / cidr-based policy, which is still actively undergoing discussion in the SIG.  In general,
+egress is expected to augment this proposal rather than modify it, with one exception (noted below).
+
+## Implementation
+
+The implmentation in Kubernetes consists of:
+- A v1beta NetworkPolicy API object
+- A field on the `Namespace` object to control isolation.
+
+### Namespace changes
+
+> Note that this section is subject to some change with the addition of egress policy as mentioned above.
+
+The `Namespace` object will be augmented with the field `networkIsolation` which will take one of the following values:
+
+When `networkIsolation=yes` is set on a namespace:
+- Pods in that Namespace will not be accessible from any other source, unless explicitly allowed by a NetworkPolicy object.
+- Pods in that Namespace will be able to access any other source (egress is uninhibited).
+
+When `networkIsolation=no` is set on a namespace:
+- Pods in that Namespace will be accessible from any other source.
+- Pods in that Namespace will be able to access any other source.
+
+The `networkIsolation` field will default to `false` if not specified.
+
+### NetworkPolicy YAML Definition and Behavior
+
+Once the namespace is isolated, a mechanism to selectively allow traffic into the namespace and between pods within
+the namespace is required.  That is accomplished through ingress rules on `NetworkPolicy`
+objects (of which there can be multiple in a single namespace).  Pods selected by
+a NetworkPolicy should allow any incoming connections that match any
+ingress rule, per the network plugin’s capabilities.
+
+```yaml
+kind: NetworkPolicy
+apiVersion: v1beta 
+metadata:
+  name:
+  namespace:
+spec:
+  podSelector:            // Standard label selector - selects pods.  
+  ingress:                // List of ingress rules (optional).             
+    - ports:              // List of allowed ports / protocols (optional).          
+        - port:           // Port on the specified protocol (optional). 
+          protocol:       // Protocol (TCP, UDP) 
```

Comment:
Regarding  the 'ports' clause - a spec like the following will allow ingress on 22,80,8080 TCP ports and UDP 6699:
    {"ports":
      [
        { "protocol": "TCP",
          "port": "8080"},
        { "protocol": "TCP",
          "port": "80"},
        { "protocol": "TCP",
          "port": "22"},
        { "protocol": "UDP",
          "port": "6699"},
      ],

This is not bed, but I find there's  a bit too much repetition in protocol names
I have experimented with a few alternatives:

    {"ports":
      [
       {"protocol": "TCP",
        "ports": ["8080", "80", "22"]
       },
       {
        "protocol": "UDP",
        "ports": ["6699"]
       }
      ]
    }

This is more compact, but the "ports" key repetition should be removed - an alternative name is needed.

    {"ports":
      [
       "TCP": ["8080", "80", "22"],
       "UDP": ["6699"]
      ] 
    } 

This is even more compact, but I don't know how do you feel about using keys which are not part a schema (I don't recall this happening anywhere in kubernetes resources)

Let me know if you reckon that there is some value in these alternative forms... I'm just throwing it here now because I think it's the right forum - but I do not feel strongly.